staking/programs/staking/src/lib.rs

```diff
@@ -721,6 +722,108 @@ pub mod staking {
+        let next_index = &mut ctx.accounts.stake_account_metadata.next_index;
+
+        let mut total_slashed = 0;
+
+        let mut i: usize = 0;
+        while i < usize::from(*next_index) {
+            msg!("i: {}", i);
+            let position = stake_account_positions.read_position(i)?;
+
+            if let Some(position_data) = position {
+                if position_data.target_with_parameters == target_with_parameters {
+                    let amount = position_data.amount;
+                    // TODO: use constants
+                    let to_slash: u64 =
+                        ((u128::from(amount) * u128::from(slash_ratio) + 1_000_000 - 1)
+                            / 1_000_000)
+                            .try_into()?;
+                    let remaining = amount - to_slash;
+                    total_slashed += to_slash;
+
+                    if remaining == 0 {
+                        stake_account_positions.make_none(i, next_index)?;
+                        continue;
+                    } else {
+                        stake_account_positions.write_position(
+                            i,
+                            &Position {
+                                amount: remaining,
+                                target_with_parameters,
+                                activation_epoch: position_data.activation_epoch,
+                                unlocking_start: position_data.unlocking_start,
+                            },
+                        )?;
+                    }
+                }
+            }
+            i += 1;
+        }
+
+        let total_amount = ctx.accounts.stake_account_custody.amount;
+        let mut governance_exposure: u64 = 0;
+
+        for i in 0..(*next_index as usize) {
+            msg!("2i: {}", i);
+            if let Some(position) = stake_account_positions.read_position(i)? {
+                if matches!(
+                    position.target_with_parameters,
+                    TargetWithParameters::Voting
+                ) {
+                    governance_exposure = governance_exposure
+                        .checked_add(position.amount)
+                        .ok_or_else(|| error!(ErrorCode::GenericOverflow))?;
+                }
+            }
+        }
+
+        if governance_exposure + total_slashed > total_amount {
+            let mut remaining = governance_exposure + total_slashed - total_amount;
+            let mut i = 0;
+            while i < usize::from(*next_index) && remaining > 0 {
+                msg!("3i: {}", i);
+                if let Some(position) = stake_account_positions.read_position(i)? {
+                    if matches!(
+                        position.target_with_parameters,
+                        TargetWithParameters::Voting
+                    ) {
+                        let amount = position.amount;
+                        let to_slash = remaining.min(amount);
+                        remaining -= to_slash;
```

Comment:
is there a more rusty way of writing all this arithmetic

Reply:
made it a little better